sbws/lib/v3bwfile.py

```diff
@@ -432,6 +435,42 @@ def add_stats(self, **kwargs):
         [setattr(self, k, str(v)) for k, v in kwargs.items()
          if k in STATS_KEYVALUES]
 
+    def add_relays_excluded_counters(self, exclusion_dict):
+        log.debug("Adding relays excluded counters.")
+        for k, v in exclusion_dict.items():
+            setattr(self, k, str(v))
+
+    def add_time_report_all_network(self):
+        """Add to the header the time it took to measure half of the network.
+
+        It is not the time the scanner actually takes on measuring all the
+        network, but the ``number_eligible_relays`` that are reported in the
+        bandwidth file.
+
+        Log also an estimated on how long it would take with the current
+        number of relays included in the bandwidth file.
+        """
+        # NOTE: in future refactor do not convert attributes to str until
+        # writing to the file.
+        elapsed_time = round(
+            (isostr_to_dt_obj(self.latest_bandwidth)
+             - isostr_to_dt_obj(self.earliest_bandwidth))
+            .total_seconds())
+
+        eligible_relays = getattr(self, 'number_eligible_relay', None)
```

Comment:
Hm, this seems to check for `number_eligible_relay` which does not exist? Is this meant to check for `number_eligible_relays`?

Does getattr work with substrings, is this broken, or something else that I don't get?
Do we need tests for this code?

In general, I'm not a fan of programming with `getattr` because it's the type of meta-programming that is really prone to errors. If it was me, I would initialize all those fields to None on the constructor, and then check that they are not None here.

Reply:
> Hm, this seems to check for number_eligible_relay which does not exist? Is this meant to check for number_eligible_relays?

yes
> Does getattr work with substrings, is this broken, or something else that I don't get?
> Do we need tests for this code?
Test would be good, but i t would prefer to do delay it after some refactoring #28684

> In general, I'm not a fan of programming with getattr because it's the type of meta-programming that is really prone to errors. If it was me, I would initialize all those fields to None on the constructor, and then check that they are not None here.

I understand your reasons.
I started to do this because we're adding new attributes progressively and when parsing old files this would make it easier to just ignore the attributes that were not present.
I think the right way to do it though would be as steam does: https://github.com/torproject/stem/blob/master/stem/descriptor/bandwidth_file.py#L61

Reply:
> Tests would be good, but i t would prefer to do delay it after some refactoring #28684
ok, i think i'll add a test